client/mysqltest.cc

```diff
@@ -3809,45 +3809,305 @@ void do_remove_files_wildcard(struct st_command *command)
+  size_t source_file_length;
+  size_t dest_file_length;
+  dest_file_length = ds_destination->length;
+  source_file_length = ds_source->length;
+  uint match_count;
+  match_count = 0;
+
+  for (uint i = 0; i < src_dir_info->number_of_files; i++) {
+    ds_source->length = source_file_length;
+    ds_destination->length = dest_file_length;
+    FILEINFO *file = src_dir_info->dir_entry + i;
+
+    /* Skip directories if recursive copy is not requested */
+    if (MY_S_ISDIR(file->mystat->st_mode) && !recursive)
+      continue;
+
+    /* Copy only those files which the pattern matches */
+    if (ds_wild->length &&
+        wild_compare(file->name, ds_wild->str, false))
+      continue;
+
+    match_count++;
+    dynstr_append(ds_source, file->name);
+    dynstr_append(ds_destination, file->name);
+
+    if (MY_S_ISDIR(file->mystat->st_mode)) {
+      DBUG_PRINT("info", ("Recursive copy files of %s to %s", ds_source->str, ds_destination->str));
+      DBUG_PRINT("info", ("listing directory: %s", ds_source->str));
+      error = recursive_copy(ds_source, ds_destination);
+    } else {
+      DBUG_PRINT("info", ("Copying file: %s to %s", ds_source->str, ds_destination->str));
+      /* MY_HOLD_ORIGINAL_MODES prevents attempts to chown the file */
+      error = (my_copy(ds_source->str, ds_destination->str, MYF(MY_HOLD_ORIGINAL_MODES)) != 0);
+    }
+
+    if (error) goto end;
+  }
+
+  /* Pattern did not match any files */
+  if (!match_count) {
+    error = 1;
+  }
+
+end:
+  my_dirend(src_dir_info);
+  my_dirend(dst_dir_info);
+  return error;
+}
+
 
 /*
   SYNOPSIS
   do_copy_file
   command	command handle
 
   DESCRIPTION
-  copy_file <from_file> <to_file>
-  Copy <from_file> to <to_file>
+  copy_file <source> <destination> [-r]
 
-  NOTE! Will fail if <to_file> exists
+  <source> can be a file, a directory, or a path with wildcard pattern
+
+  - When <source> is a directory, '-r' must be given to copy recursively
+  - When <source> is a wildcard pattern, '-r' is optional. If given, then
+    directories that match the pattern will also be copied.
 */
 
 void do_copy_file(struct st_command *command)
 {
   int error;
-  static DYNAMIC_STRING ds_from_file;
-  static DYNAMIC_STRING ds_to_file;
+  static DYNAMIC_STRING ds_source;
+  static DYNAMIC_STRING ds_destination;
+  static DYNAMIC_STRING ds_wild;
+  static DYNAMIC_STRING ds_recursive;
+  static bool recursive;
+  MY_STAT stat_buff;
   const struct command_arg copy_file_args[] = {
-    { "from_file", ARG_STRING, TRUE, &ds_from_file, "Filename to copy from" },
-    { "to_file", ARG_STRING, TRUE, &ds_to_file, "Filename to copy to" }
+    { "source", ARG_STRING, TRUE, &ds_source, "File/dir name to copy from" },
+    { "destination", ARG_STRING, TRUE, &ds_destination, "File/dir name to copy to" },
+    { "recursive", ARG_STRING, FALSE, &ds_recursive, "Optional argument indicating if directories should be copied recursively" }
   };
-  DBUG_ENTER("do_copy_file");
```

Comment:
I believe removing this is the cause, you can't run `DBUG_VOID_RETURN` without a `DBUG_ENTER`.

Reply:
Thanks @LinuxJedi! I fixed it and now it looks good.